### PR TITLE
[JavaScript] Improved identifier and number lexing.

### PR DESCRIPTION
**Pull request by Thom1729**
Implemented the identifier rules from the spec.

The syntax previously used `\b` in a lot of rules. This is not quite right, and it leads to some bugs:

```js
function$NotAFunction
```

Where `\b` followed identifier characters, I replaced it with a negative lookahead. (Testing showed that there was no effect on performance.) Where `\b` *preceded* identifier characters, we would need a lookbehind, which the fast regexp engine does not support. However, we don't actually need assertions at the beginning of matches if we maintain the invariant that the "cursor" should never pause with identifier characters on both sides. This is guaranteed if every match that ends with an identifier character ends with the negative lookahead.

Implementing this correctly required changing the implementation of numeric literals. I added more tests and improved the behavior of numbers in some corner cases.

### CODE REVIEW

**Comment by michaelblyons** [comment]
Ruby [may need](https://github.com/sublimehq/Packages/issues/1324#issuecomment-369667473) this sort of thing, too.

File: JavaScript/JavaScript.sublime-syntax [hunk 1]
```diff
@@ -8,12 +8,16 @@ file_extensions:
...
-  func_lookahead: '\s*\b(async\s+)?function\b'
-  arrow_func_lookahead: '\s*(\basync\s*)?([_$[:alpha:]][_$[:alnum:]]*|\(([^()]|\([^()]*\))*\))\s*=>'
+  identifier_start: '[_$\p{L}\p{Nl}]'
+  identifier_part: '[_$\p{L}\p{Nl}\p{Mn}\p{Mc}\p{Nd}\p{Pc}\x{200C}\x{200D}]'
+  identifier_break: (?!{{identifier_part}})
+
+  identifier: '{{identifier_start}}{{identifier_part}}*{{identifier_break}}'
+  constant_identifier: '[[:upper:]]{{identifier_part}}*{{identifier_break}}'
+  dollar_only_identifier: '\${{identifier_break}}'
+  dollar_identifier: '(\$){{identifier_part}}*{{identifier_break}}+'
```
Comment (sharkdp) [inline JavaScript/JavaScript.sublime-syntax]: @Thom1729 This `dollar_identifier` regular expression seems to cause errors when the JavaScript language definition is used within [`syntect`](https://github.com/trishume/syntect/).

In particular, it uses `+` on `{{identifier_break}}`, which is a negative-lookahead group: `(?!...)+`. Is this valid regex syntax? Could/should the `+` be removed?

Comment (Thom1729) [reply JavaScript/JavaScript.sublime-syntax]: It's valid but redundant. I don't know why syntect doesn't handle this, but if you open up a PR to remove the `+` I'll endorse it.

Comment (robinst) [reply JavaScript/JavaScript.sublime-syntax]: (for anyone arriving here, @sharkdp raised a PR: https://github.com/sublimehq/Packages/pull/1525)
